### PR TITLE
Improved result/error handling during migration

### PR DESCRIPTION
**Pull request by grigoryk**
Best reviewed commit by commit.

Changes should be mostly non-functional - just adding more coverage for exception handling.

In one narrow case we were counting potential failures as successes, swallowing the problem. The problem would still get reported as a caught exception via Sentry, and we haven't seen it happen.

---

### Pull Request checklist

- [ ] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/android-components/blob/master/docs/changelog.md) or does not need one
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### CODE REVIEW

**Comment by grigoryk** [comment]
Did a few test runs locally, and this does appear to work as expected.

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt [hunk 1]
```diff
@@ -734,23 +733,23 @@ class FennecMigrator private constructor(
...
         // See application-services repository: https://github.com/mozilla/application-services/commit/b2e2edcc06a04503d493e1733b0d566815feac7c#diff-216f62325632ae6549587b038b21cfe0
-        try {
+        return try {
             MigrationBookmarks.detected.add(migrationMetrics.getInt("num_total"))
             MigrationBookmarks.migrated["succeeded"].add(migrationMetrics.getInt("num_succeeded"))
             MigrationBookmarks.migrated["failed"].add(migrationMetrics.getInt("num_failed"))
             // Assuming that 'total_duration' is in milliseconds.
             MigrationBookmarks.duration.setRawNanos(migrationMetrics.getLong("total_duration") * 1000000)
+
+            logger.debug("Migrated history.")
```
Comment (csadilek) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:743]: This is bookmarks, not history.

**Comment by grigoryk** [comment]
Thank you, @csadilek !

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt [hunk 2]
```diff
@@ -597,22 +603,27 @@ class FennecMigrator private constructor(
...
+                    Migration.Addons -> migrateAddons()
+                    Migration.TelemetryIdentifiers -> migrateTelemetryIdentifiers()
+                    Migration.SearchEngine -> migrateSearchEngine()
+                    Migration.PinnedSites -> migratePinnedSites()
+                }
+            } catch (e: Exception) {
+                logger.error("Unexpected error while migrating $versionedMigration", e)
+                Result.Failure<Any>(e)
+            } finally {
+                versionedMigration.migration.metricTotalDuration().stop()
```
Comment (pocmo) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:624]: Is it in any way problematic for data analysis if we include incomplete migrations into the duration measurement?

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt [hunk 3]
```diff
@@ -597,22 +603,27 @@ class FennecMigrator private constructor(
...
             }
 
-            versionedMigration.migration.metricTotalDuration().stop()
```
Comment (pocmo) [inline components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:614]: If we measured total duration, did we see this drop off consistently with the funnel?

**Comment by pocmo** [comment]
Did some migrations with this branch and it looks good.

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt [hunk 4]
```diff
@@ -597,22 +603,27 @@ class FennecMigrator private constructor(
...
+                    Migration.Logins -> migrateLogins()
+                    Migration.Settings -> migrateSharedPrefs()
+                    Migration.Addons -> migrateAddons()
+                    Migration.TelemetryIdentifiers -> migrateTelemetryIdentifiers()
+                    Migration.SearchEngine -> migrateSearchEngine()
+                    Migration.PinnedSites -> migratePinnedSites()
+                }
+            } catch (e: Exception) {
+                logger.error("Unexpected error while migrating $versionedMigration", e)
+                Result.Failure<Any>(e)
```
[the remark this answers is not in the document]
Comment (grigoryk) [reply components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:622]: Ah, no, that's an omission.

File: components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt (hunk 2 above)
Comment (grigoryk) [reply components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:624]: I think this should be okay, but let's confirm with @fbertsch ?

Comment (grigoryk) [reply components/support/migration/src/main/java/mozilla/components/support/migration/FennecMigrator.kt:624]: We will now include migration duration even for the completely broken migrations. Before, it'll be only for the gracefully failed/successful ones.

**Comment by grigoryk** [comment]
bors r=csadilek,pocmo

**Comment by grigoryk** [comment]
Re-ran this with deliberate crashes sprinkled in within the migrations, everything's behaving correctly.

**Comment by grigoryk** [comment]
bors r-

going to squash first.

**Comment by grigoryk** [comment]
bors r=csadilek,pocmo

**Pull request merged**